### PR TITLE
fix: unrug app when missing chain adapter

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Types things what they actually are at runtime.

- Removes non-null assertions, as they are invalid
- Makes `AssetsById` a `PartialRecord`
- Handles all cases where an `Asset` lookup by dynamic key can be `undefined`

Unrugs local state when Osmosis (or any other) ChainAdapter is missing and there are accounts available.
This might save us pain down the road.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Medium. Touches a lot of thing - we'll want a good old regression check.

## Testing

In most cases the app should function exactly as it did before.

### Engineering

Is Osmosis was enabled, and then turned off, we should no longer have an unusable app.

### Operations

☝️ 

## Screenshots (if applicable)

N/A

### CODE REVIEW

**Comment by 0xApotheosis** [comment]
As a follow-up tidying PR, I'm thinking of replacing all of the throwing logic with an Asset getter helper that throws if an asset is not found.

**Review by gomesalexandre — COMMENTED** [review]
Comments - first pass, discussed this one in-person and conceptually agree from a very high-level, however my main concern with this is throwing straight from the render methods, technically speaking that's fine, though that's not a very react-y way to do things and I've never seen this in the wild.

Not such a big deal to go against conventions, but a consequence of throwing vs. simply returning null is that instead of having a "working", empty state you can debug, you'll now have the app fully error because of the error boundary.

We should:

1. find a way to have more local error boundaries, maybe with some composition of sorts so that only that one component
2. make sure we don't throw when we shouldn't, undefined xyz are expected when data is loading e.g deep linking, while data is populating, empty states are expected and are not an exception. Definitely not as big of an issue not that we have persistence, but still, I'd be very careful considering falsy values exceptions in the context of components

File: src/components/AssetAccounts/AssetAccountRow.tsx [hunk 1]
```diff
@@ -50,7 +50,7 @@ export const AssetAccountRow = ({
   const rowAssetId = assetId ? assetId : feeAssetId
   const asset = useAppSelector(state => selectAssetById(state, rowAssetId ?? ''))
   const feeAsset = useAppSelector(state => selectAssetById(state, feeAssetId ?? ''))
-  const { assetReference, assetNamespace } = fromAssetId(asset.assetId)
+  const { assetReference, assetNamespace } = (asset && fromAssetId(asset.assetId)) ?? {}
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/AssetAccounts/AssetAccountRow.tsx:53]: I'm assuming you mean something like `const { assetReference, assetNamespace } = asset ? fromAssetId(asset.assetId) : {}`, which does not work for object destructuring of a potentially `undefined` object.

File: src/components/TransactionHistory/AssetTransactionHistory.tsx [hunk 2]
```diff
@@ -37,7 +37,8 @@ export const AssetTransactionHistory: React.FC<AssetTransactionHistoryProps> = (
   } = useWallet()
 
   const asset = useAppSelector(state => selectAssetById(state, assetId ?? ''))
-  const chainId = asset.chainId
+  const chainId: ChainId = asset?.chainId ?? ''
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/TransactionHistory/AssetTransactionHistory.tsx:40]: Indeed - explicitly typing (distinct from casting) a constant as `ChainId` narrows any string literal to `ChainId`.

File: src/components/AssetAccounts/AssetAccountRow.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/AssetAccounts/AssetAccountRow.tsx:53]: Yeah probably safer that way - having `asset` effectively means that we have a valid assetId which means `fromAssetId()` would return an object, but let's be safu

File: src/features/defi/providers/cosmos/components/CosmosManager/Claim/components/Confirm.tsx [hunk 3]
```diff
@@ -55,9 +58,12 @@ export const Confirm: React.FC<ConfirmProps> = ({ accountId, onNext }) => {
     assetNamespace: 'slip44',
     assetReference,
   })
-  const feeAsset = useAppSelector(state => selectAssetById(state, feeAssetId))
+  const feeAsset = useAppSelector(state => selectAssetById(state, opportunity?.assetId))
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/features/defi/providers/cosmos/components/CosmosManager/Claim/components/Confirm.tsx]: Excellent catch sir, this was a not an intended change.

Fixed in https://github.com/shapeshift/web/pull/3501/commits/84d8e5f0ff65f11eaabbba9139fcc9ec8359c4ae

File: src/components/Modals/FiatRamps/views/Overview.tsx [hunk 4]
```diff
@@ -222,7 +222,9 @@ export const Overview: React.FC<OverviewProps> = ({
     ))
   }, [])
 
-  return (
+  const asset = useMemo(() => assetsById[assetId], [assetId, assetsById])
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Modals/FiatRamps/views/Overview.tsx]: https://github.com/shapeshift/web/pull/3501/commits/84d8e5f0ff65f11eaabbba9139fcc9ec8359c4ae

File: src/components/AssetIcon.tsx [hunk 5]
```diff
@@ -23,13 +23,13 @@ type AssetWithNetworkProps = {
 const AssetWithNetwork: React.FC<AssetWithNetworkProps> = ({ assetId, icon, ...rest }) => {
   const asset = useAppSelector(state => selectAssetById(state, assetId ?? ''))
   const feeAsset = useAppSelector(state => selectFeeAssetById(state, assetId))
-  const showFeeAsset = asset.assetId !== feeAsset.assetId
+  const showFeeAsset = asset?.assetId !== feeAsset?.assetId
   const boxShadow = useColorModeValue(
-    `0 0 0 0.2em ${feeAsset.color}35, 0 0 0.5em 2px rgba(255,255,255,.5)`,
-    `0 0 0 0.2em ${feeAsset.color}50, 0 0 0.5em 2px rgba(0,0,0,.5)`,
+    `0 0 0 0.2em ${feeAsset?.color}35, 0 0 0.5em 2px rgba(255,255,255,.5)`,
+    `0 0 0 0.2em ${feeAsset?.color}50, 0 0 0.5em 2px rgba(0,0,0,.5)`,
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/AssetIcon.tsx]: https://github.com/shapeshift/web/pull/3501/commits/84d8e5f0ff65f11eaabbba9139fcc9ec8359c4ae

File: src/state/slices/opportunitiesSlice/selectors.ts [hunk 6]
```diff
@@ -310,15 +310,17 @@ export const selectAggregatedEarnUserStakingOpportunities = createDeepEqualOutpu
...
-            .times(marketData[asset?.assetId ?? underlyingAsset?.assetId]?.price ?? '0')
-            .div(bn(10).pow(asset?.precision ?? underlyingAsset?.precision))
+            .times(marketData[asset?.assetId ?? underlyingAsset?.assetId ?? '']?.price ?? '0')
+            .div(bn(10).pow(asset?.precision ?? underlyingAsset?.precision ?? 1))
             .toString(),
           isLoaded: true,
-          icons: opportunity.underlyingAssetIds.map(assetId => assets[assetId].icon),
+          icons: opportunity.underlyingAssetIds
+            .map(assetId => assets[assetId]?.icon)
+            .filter(isSome),
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/slices/opportunitiesSlice/selectors.ts]: Fixed in https://github.com/shapeshift/web/pull/3501/commits/ecd65727582f5066cfaf04654326bbd57c482bbc

File: src/state/slices/opportunitiesSlice/selectors.ts [hunk 7]
```diff
@@ -270,7 +270,7 @@ export const selectAggregatedEarnUserStakingOpportunityByStakingId = createDeepE
         .times(marketData[opportunity.underlyingAssetId as AssetId]?.price ?? '0')
         .toString(),
       isLoaded: true,
-      icons: opportunity.underlyingAssetIds.map(assetId => assets[assetId].icon),
+      icons: opportunity.underlyingAssetIds.map(assetId => assets[assetId]?.icon).filter(isSome),
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/slices/opportunitiesSlice/selectors.ts]: https://github.com/shapeshift/web/pull/3501/commits/ecd65727582f5066cfaf04654326bbd57c482bbc

File: src/features/defi/providers/fox-farming/components/FoxFarmingManager/Overview/FoxFarmingOverview.tsx [hunk 8]
```diff
@@ -78,8 +81,10 @@ export const FoxFarmingOverview: React.FC<FoxFarmingOverviewProps> = ({
     selectUserStakingOpportunityByUserStakingId(state, opportunityDataFilter),
   )
 
-  const underlyingAssetsIcons = useMemo(
-    () => opportunityData?.underlyingAssetIds.map(assetId => assets[assetId].icon),
+  const underlyingAssetsIcons: string[] = useMemo(
+    () =>
+      opportunityData?.underlyingAssetIds.map(assetId => assets[assetId]?.icon).filter(isSome) ??
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/features/defi/providers/fox-farming/components/FoxFarmingManager/Overview/FoxFarmingOverview.tsx]: Fixed in https://github.com/shapeshift/web/pull/3501/commits/ecd65727582f5066cfaf04654326bbd57c482bbc

File: src/state/slices/opportunitiesSlice/resolvers/idle/index.ts [hunk 9]
```diff
@@ -107,7 +107,7 @@ export const idleStakingOpportunitiesMetadataResolver = async ({
           },
           // Idle opportunities wrap a single yield-bearing asset, so the ratio will always be 1
           underlyingAssetRatios: ['1'],
-          name: `${underlyingAsset.symbol} Vault`,
+          name: `${underlyingAsset?.symbol} Vault`,
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/slices/opportunitiesSlice/resolvers/idle/index.ts]: https://github.com/shapeshift/web/pull/3501/commits/52fe1f1851d439cb2c911a5fe6e37ae1d3029a6e

**Comment by 0xApotheosis** [comment]
@gomesalexandre should be good to go, ser.

File: src/features/defi/providers/fox-farming/components/FoxFarmingManager/Overview/FoxFarmingOverview.tsx [hunk 10]
```diff
@@ -83,8 +82,9 @@ export const FoxFarmingOverview: React.FC<FoxFarmingOverviewProps> = ({
 
   const underlyingAssetsIcons: string[] = useMemo(
     () =>
-      opportunityData?.underlyingAssetIds.map(assetId => assets[assetId]?.icon).filter(isSome) ??
-      [],
+      opportunityData?.underlyingAssetIds
+        .map(assetId => assets[assetId]?.icon)
+        .map(icon => icon ?? '') ?? [],
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/components/FoxFarmingManager/Overview/FoxFarmingOverview.tsx:87]: 🧠

**Review by gomesalexandre — APPROVED** [review]
Confirmed Osmosis flags toggling isn't rugging the app anymore.

Discussed this one out and let's go slightly against convention with render method level throwing for the time being.

**Review by 0xdef1cafe — APPROVED** [review]
similarly LGTM

rationale - throwing in a view will be caught by the error boundary, which while it might seem like an anti-pattern, is categorically better than crashing the app

**Pull request merged**